@xen-orchestra/backups-cli/index.js

```
@@ -0,0 +1,354 @@
+const { createSyntheticStream, default: Vhd, mergeVhd } = require('vhd-lib')
+const { curryRight, flatten } = require('lodash')
+const { dirname, resolve } = require('path')
+const { DISK_TYPE_DIFFERENCING } = require('vhd-lib/dist/_constants')
+const { pipe, promisifyAll } = require('promise-toolbox')
+
+const fs = promisifyAll(require('fs'))
+const handler = require('@xen-orchestra/fs').getHandler({ url: 'file://' })
+
+// -----------------------------------------------------------------------------
+
+const asyncMap = curryRight((iterable, fn) =>
+  Promise.all(
+    Array.isArray(iterable) ? iterable.map(fn) : Array.from(iterable, fn)
+  )
+)
+
+const filter = (...args) => thisArg => thisArg.filter(...args)
+
+// TODO: better check?
+
+// our heuristic is not good enough, there has been some false positives
+// (detected as invalid by us but valid by `tar` and imported with success),
+// either:
+// - these files were normal but the check is incorrect
+// - these files were invalid but without data loss
+// - these files were invalid but with silent data loss
+//
+// maybe reading the end of the file looking for a file named
+// /^Ref:\d+/\d+\.checksum$/ and then validating the tar structure from it
+//
+// https://github.com/npm/node-tar/issues/234#issuecomment-538190295
+const isValidTar = async path => {
+  const fd = await fs.open(path, 'r')
+  try {
+    const { size } = await fs.fstat(fd)
+    if (size <= 1024 || size % 512 !== 0) {
+      return false
+    }
+
+    const buf = Buffer.allocUnsafe(1024)
+    assert.strictEqual(
+      await fs.read(fd, buf, 0, buf.length, size - 1 - buf.length),
+      buf.length
+    )
+    return buf.every(_ => _ === 0)
+  } finally {
+    fs.close(fd).catch(noop)
+  }
+}
+
+const noop = Function.prototype
+
+const readDir = path =>
+  fs.readdir(path).then(
+    entries => {
+      entries.forEach((entry, i) => {
+        entries[i] = `${path}/${entry}`
+      })
+
+      return entries
+    },
+    () => []
```

Why?

---

I don't want my code to fail and I want to keep as simple as possible, simply returning `[]` is enough.

---

But then it would make you think that everything is fine by failing silently.